File: tisca/R/estimands.R
# ----------------------------------------------------------------------------
# TISCA v2 -- estimands.R
# Per-replication estimands from the reference estimand table (docs/estimand_table.md,
# P1-T1 rev. 2) and the E3 per-replication metric columns. This module turns raw
# per-unit or per-replication data into the replicate-level loss L_j on which
# paired contrasts are built.
#
# Rows implemented here map to the estimand table as follows (nominal level
# parameterised per Section 3.2):
#   1  PEHE (rooted):            pehe(tau_hat, tau)
#   2  CATE MSE (unrooted):      cate_mse(tau_hat, tau)
#   3a ATE squared error:        ate_sq_err(ate_hat, ate_true)
#   3b RMSE_ATE estimand + MCSE: rmse_ate(mcse = ...)  (derived; see row 3b)
#   4  ATE coverage indicator:   ate_cov(ate_hat, ate_true, ci_lo, ci_hi)
#   5  CATE coverage proportion: cate_cov(tau_hat, tau, ci_lo_i, ci_hi_i)
#   6  CIL (width):              cil(ci_lo, ci_hi)
#   7  Bias (CATE):              cate_bias(tau_hat, tau)
#   8  Calibration deviation flow is handled at analysis time (apply abs to the
#      estimate, NEVER per replication); see Section 3.1 of the estimand table.
#   9  Interval score:           interval_score(x, lo, hi, c)
#   10 CRPS:                     crps(...)  -- supplied externally (scoringRules);
#      this module only carries the column as a loss L_j.
#   11 Runtime:                  passed through as a replicate-level column.
#   12 Convergence/failure flag: passed through as a binary column.
# ----------------------------------------------------------------------------

## ---- unit-level / replicate-level estimators --------------------------------

#' Rooted PEHE: PEHE_j = sqrt(mean_i (tau_hat_{j,i} - tau_{j,i})^2).
#' @param tau_hat vector of per-unit CATE estimates on one replication
#' @param tau true per-unit CATE on the same replication
#' @return scalar rooted PEHE for the replication
pehe <- function(tau_hat, tau) {
  sqrt(mean((tau_hat - tau)^2, na.rm = FALSE))
}

#' Unrooted CATE MSE: Q_j = mean_i (tau_hat_{j,i} - tau_{j,i})^2.
#' @inheritParams pehe
#' @return scalar unrooted MSE for the replication
cate_mse <- function(tau_hat, tau) {
  mean((tau_hat - tau)^2, na.rm = FALSE)
}

#' ATE squared error: Q_{ATE,j} = (ate_hat - ate_true)^2.
#' @param ate_hat estimated ATE for the replication
#' @param ate_true true ATE for the replication
#' @return scalar squared error
ate_sq_err <- function(ate_hat, ate_true) {
  (ate_hat - ate_true)^2
}

#' RMSE_ATE estimand is sqrt(E[Q_ATE]); this computes the estimator and its
#' delta-method MCSE from the raw per-replication squared errors (row 3b).
#'
#' The MCSE on the rooted scale is s_Q / (2*sqrt(J)*sqrt(mean(Q))). This is the
#' corrected formula; dropping the inner root inflates/understates by 2x-5x in the
#' PEHE-scale regime (estimand table Section 3, row 3b, verified numerically).
#' The recommended contrast path is to pair-test on the squared scale (row 3a)
#' and re-root the estimate, so this routine is provided for the plain report.
#'
#' @param q_ate vector of per-replication squared ATE errors
#' @return list(rmse = sqrt(mean(q)), mcse = delta-method MCSE)
rmse_ate <- function(q_ate) {
  mq <- mean(q_ate)
  if (length(q_ate) < 2) stop("need >= 2 replications for an MCSE")
  s <- stats::sd(q_ate)
  list(
    rmse  = sqrt(mq),
    mcse  = if (mq > 0) s / (2 * sqrt(length(q_ate)) * sqrt(mq)) else 0
  )
}

#' ATE coverage indicator (row 4): 1(ate_true in [lo, hi]).
#' @param ate_true true ATE
#' @param ci_lo lower credible/CI bound on ATE
#' @param ci_hi upper credible/CI bound on ATE
#' @return 0/1 indicator
ate_cov <- function(ate_true, ci_lo, ci_hi) {
  as.integer(ate_true >= ci_lo & ate_true <= ci_hi)
}

#' CATE coverage proportion (row 5): mean_i 1(tau_i in [lo_i, hi_i]).
#' The inputs are parallel per-unit vectors; a correctly calibrated midpoint of
#' the interval is not assumed, the bounds are used as given.
#' @param tau_i true per-unit CATE vector
#' @param ci_lo_i per-unit lower bound
#' @param ci_hi_i per-unit upper bound
#' @return scalar proportion covered in [0,1]
cate_cov <- function(tau_i, ci_lo_i, ci_hi_i) {
  n <- length(tau_i)
  if (n == 0) return(NA_real_)
  mean(tau_i >= ci_lo_i & tau_i <= ci_hi_i, na.rm = FALSE)
}

#' Mean credible/CI width (row 6): mean_i (hi_i - lo_i) [CATE], or (hi-lo) [ATE].
#' @param ci_lo numeric vector (per-unit or scalar)
#' @param ci_hi numeric vector (per-unit or scalar)
#' @return scalar mean width
cil <- function(ci_lo, ci_hi) {
  mean(ci_hi - ci_lo, na.rm = FALSE)
}

#' CATE bias (row 7): mean_i (tau_hat_i - tau_i).
#' @param tau_hat per-unit CATE estimates
#' @param tau per-unit true CATEs
#' @return scalar bias
cate_bias <- function(tau_hat, tau) {
  mean(tau_hat - tau, na.rm = FALSE)
}

#' Interval score (row 9), Winkler form at nominal level 1-c.
#' IS = (u-l) + (2/c)(l - x)1(x<l) + (2/c)(x - u)1(x>u).
#' Vectorised over units; returns per-unit vector. Lower is better.
#' @param x observed value (may be a vector)
#' @param lo lower bound (vector)
#' @param hi upper bound (vector)
#' @param c nominal miscoverage of the interval (0.05 for the 95% level, 0.50 for the 50% level)
#' @return numeric vector of per-unit interval scores
interval_score <- function(x, lo, hi, c = 0.05) {
  term <- (hi - lo) +
          (2 / c) * (lo - x) * (x < lo) +
          (2 / c) * (x - hi) * (x > hi)
  term
}

#' Mean interval score over units (the scalar D-loss the tables use).
#' @param x observed value (vector)
#' @param lo lower bound (vector)
#' @param hi upper bound (vector)
#' @param c nominal miscoverage
#' @return scalar mean interval score for the replication
mean_interval_score <- function(x, lo, hi, c = 0.05) {
  mean(interval_score(x, lo, hi, c = c), na.rm = FALSE)
}

## ---- validation helpers (used by validate.R) --------------------------------

#' Check a candidate L_j vector is a valid per-replication loss for a contrast.
#' @param vec candidate replicate-level column
#' @return TRUE if length >= 2 and all finite
is_perrep_loss <- function(vec) {
  length(vec) >= 2 && all(is.finite(vec))
}

#' Check a column is a valid binary (0/1) replicate-level indicator (rows 4, 12).
#' @param vec candidate indicator column
#' @return TRUE if all values in {0,1} and length >= 1
is_binary_indicator <- function(vec) {
  length(vec) >= 1 && all(vec %in% c(0, 1)) && all(is.finite(vec))
}

#' Check a bounded proportion column (row 5), 0 <= L_j <= 1.
#' @param vec candidate proportion column
#' @return TRUE if all values in [0,1] and length >= 2
is_bounded_proportion <- function(vec) {
  length(vec) >= 2 && all(is.finite(vec)) &&
    all(vec >= 0 & vec <= 1)
}

File: tisca/python/tisca/estimands.py
"""Estimand table rows — `estimands.py`.

Implements the replicate-level losses, across-replication estimators, and MCSE
formulas of `docs/estimand_table.md` row by row. Every function here operates on
a **per-replication** ``L_j`` (or on per-unit quantities aggregated to one value
per replication *before* any across-replication statistic).

Conventions (estimand table §1, all verbatim):
* ``i`` = unit index within a replication, ``j`` = replication index.
* ``tau_i`` = unit-level CATE; ``tau_j`` = replication-level ATE.
* rooted PEHE is ``E[sqrt(Q)]``; unrooted CATE-MSE is ``E[Q]``; these differ by
  Jensen (``E[sqrt Q] != sqrt(E[Q])``).
* RMSE_ATE (row 3b) is ``sqrt(E[Q_ATE])`` only, never a re-rooted per-replication
  value.
"""

from __future__ import annotations

import numpy as np

__all__ = [
    "rooted_pehe",
    "cate_mse",
    "ate_squared_error",
    "rmse_ate",
    "rmse_ate_mcse_delta",
    "ate_coverage",
    "cate_coverage",
    "interval_score_unit",
    "interval_score",
    "crps_mean",
    "calibration_deviation",
    "mean_mcse",
    "mean_estimate",
    "nc_ci",
    "winkler_score",
]


def rooted_pehe(tau_hat, tau_true, *, axis=-1) -> np.ndarray:
    """Row 1: rooted PEHE, ``PEHE_j = sqrt(mean_i (tau_hat - tau_true)^2)``.

    ``tau_hat``, ``tau_true`` may be per-unit vectors (returns one scalar per cell
    along ``axis``) or already per-replication columns (returns ``sqrt(mean(..))``
    of the single squared value).
    """
    tau_hat = np.asarray(tau_hat, dtype=float)
    tau_true = np.asarray(tau_true, dtype=float)
    q = np.mean((tau_hat - tau_true) ** 2, axis=axis)
    return np.sqrt(q)


def cate_mse(tau_hat, tau_true, *, axis=-1) -> np.ndarray:
    """Row 2: unrooted CATE MSE, ``Q_j = mean_i (tau_hat - tau_true)^2``."""
    tau_hat = np.asarray(tau_hat, dtype=float)
    tau_true = np.asarray(tau_true, dtype=float)
    return np.mean((tau_hat - tau_true) ** 2, axis=axis)


def ate_squared_error(ate_hat, ate_true) -> np.ndarray:
    """Row 3a: per-replication ATE squared error ``Q_ATE,j = (tau_hat - tau)^2``."""
    ate_hat = np.asarray(ate_hat, dtype=float)
    ate_true = np.asarray(ate_true, dtype=float)
    return (ate_hat - ate_true) ** 2


def rmse_ate(sq_err) -> float:
    """Row 3b: ``RMSE_ATE = sqrt( E[Q_ATE] )``.

    Operates on the per-replication squared errors ``Q_ATE,j`` (row 3a). The
    square root is applied to the *averaged* value, never to a single
    replication. This is the estimand that exists; "per-replication RMSE" does
    not (rejected in ``validate.py``).
    """
    q = np.asarray(sq_err, dtype=float)
    return float(np.sqrt(np.mean(q)))


def rmse_ate_mcse_delta(q_sq_err, *, sqrt_n: bool = True) -> float:
    """Row 3b delta-method MCSE for RMSE_ATE.

    With ``g(x)=sqrt(x)``, ``g'(x)=1/(2*sqrt(x))`` the delta-method MCSE is
    ``MCSE ≈ s_{Q} / ( 2*sqrt(J) * sqrt(mean(Q)) )``. The denominator carries
    ``sqrt(mean(Q))``, **not** ``mean(Q)`` (see estimand table row 3b).
    Returning the delta-method cross-check; the paired bootstrap is preferred.
    """
    q = np.asarray(q_sq_err, dtype=float)
    mu = float(np.mean(q))
    s = float(np.std(q, ddof=1))
    J = float(q.size)
    denom = 2.0 * np.sqrt(J) * np.sqrt(mu)
    out = s / denom if sqrt_n else s / (2.0 * mu)
    return float(out)


def ate_coverage(ate_hat, ci_lo, ci_hi, ate_true) -> np.ndarray:
    """Row 4: ATE coverage indicator ``1( tau_j in CI_j )``, binary per replication."""
    a = np.asarray(ate_hat, dtype=float)
    lo = np.asarray(ci_lo, dtype=float)
    hi = np.asarray(ci_hi, dtype=float)
    t = np.asarray(ate_true, dtype=float)
    return ((lo <= t) & (t <= hi)).astype(float)


def cate_coverage(tau_hat, ci_lo, ci_hi, tau_true, *, axis=-1) -> np.ndarray:
    """Row 5: CATE coverage proportion ``mean_i 1(tau_i in CI_i)`` per replication."""
    tau_hat = np.asarray(tau_hat, dtype=float)
    lo = np.asarray(ci_lo, dtype=float)
    hi = np.asarray(ci_hi, dtype=float)
    tau_true = np.asarray(tau_true, dtype=float)
    return np.mean((lo <= tau_true) & (tau_true <= hi), axis=axis).astype(float)


def interval_score_unit(lo, hi, x, *, level) -> float:
    """Row 9: Gneiting-Raftery interval score for one ``(lo, hi, x)`` triplet.

    For a ``(1-c)`` central interval, ``c = 1 - level`` and
    ``IS = (hi - lo) + (2/c)(lo - x) 1[x<lo] + (2/c)(x - hi) 1[x>hi]``.
    Lower is better (a proper scoring rule jointly penalising width and
    non-coverage).
    """
    c = 1.0 - level
    if not 0.0 < level < 1.0:
        raise ValueError("level must be in (0, 1).")
    lo = float(lo)
    hi = float(hi)
    x = float(x)
    pen_lo = (2.0 / c) * (lo - x) if x < lo else 0.0
    pen_hi = (2.0 / c) * (x - hi) if x > hi else 0.0
    return (hi - lo) + pen_lo + pen_hi


def interval_score(lo, hi, x, *, level, axis=-1) -> np.ndarray:
    """Row 9: mean interval score over units ``IS_j = mean_i IS(...)``.

    Vectorised over arrays ``lo, hi, x`` (per-unit lower/upper/observed).
    """
    level = float(level)
    lo = np.asarray(lo, dtype=float)
    hi = np.asarray(hi, dtype=float)
    x = np.asarray(x, dtype=float)
    c = 1.0 - level
    width = hi - lo
    pen_lo = (2.0 / c) * (lo - x) * (x < lo)
    pen_hi = (2.0 / c) * (x - hi) * (x > hi)
    return np.mean(width + pen_lo + pen_hi, axis=axis)


#: Alias consistent with the manuscript/Winkler literature.
winkler_score = interval_score


def crps_mean(unit_crps) -> np.ndarray:
    """Row 10: per-replication CRPS ``CRPS_j = mean_i CRPS_i``.

    ``unit_crps`` holds the per-unit CRPS values; the mean over units is the
    replication-level loss (the ``scoringRules`` values already live on this
    unit scale in the case-study CSVs).
    """
    u = np.asarray(unit_crps, dtype=float)
    return np.mean(u, axis=-1)


def calibration_deviation(cov_mean, nominal: float) -> float:
    """Row 8: calibration deviation applied to the *estimate*, never per replication.

    ``|mean(Cov_j) - nominal|``. The estimand table (§3.1) forbids
    ``E[|Cov_j - nominal|]`` because it confounds miscalibration with
    replication-level dispersion. The deviation is formed **after** averaging.
    """
    return float(abs(float(cov_mean) - float(nominal)))


def _std(x, ddof: int = 1) -> float:
    x = np.asarray(x, dtype=float)
    if x.size < 2:
        return 0.0
    return float(np.std(x, ddof=ddof))


def mean_estimate(values) -> float:
    """Across-replication estimator ``theta_hat = mean(L_j)`` (estimand table §2)."""
    return float(np.mean(np.asarray(values, dtype=float)))


def mean_mcse(values) -> float:
    """MCSE of the sample mean: ``s_L / sqrt(J)`` (estimand table §2)."""
    v = np.asarray(values, dtype=float)
    if v.size < 2:
        return float("nan")
    return _std(v, ddof=1) / np.sqrt(v.size)


def nc_ci(mean, se, *, level: float, df: int) -> tuple[float, float]:
    """Normal-approximation / t CI ``mean +/- t_{1-alpha/2, J-1} se``.

    Uses the Student-t quantile with ``df = J - 1`` degrees of freedom (the
    discrete small-``J`` correction over the plain normal CI).
    """
    from scipy.stats import t as _t

    if not 0.0 < level < 1.0:
        raise ValueError("level must be in (0, 1).")
    crit = float(_t.ppf((1.0 + level) / 2.0, df=df))
    return (float(mean) - crit * float(se), float(mean) + crit * float(se))


def _binomial_mcse(p_hat, J: int) -> float:
    """MCSE of a proportion: ``sqrt(p_hat (1 - p_hat) / J)`` (rows 4, 12)."""
    p = float(p_hat)
    return float(np.sqrt(p * (1.0 - p) / J)) if J > 0 else float("nan")
